### ISSUE TITLE
Add ability to force Rule and Subscription shortener

### ISSUE THREAD

**Issue by dv42** [Particular/NServiceBus.Transport.AzureServiceBus#39]
**Description**
Breaking change was introduced between NServiceBus.Azure.Transports.WindowsAzureServiceBus and NServiceBus.Transport.AzureServiceBus when ISanitizationStrategy was removed in case if it was altering names even if they were less than 50 symbols

**To Reproduce**
1. Create endpoint configuration with name "Alpha.Beta.Gamma.Endpoint" using NServiceBus.Azure.Transports.WindowsAzureServiceBus
2. Configure ISanitizationStrategy to alter subscription name to "A.B.G.Endpoint"
3. Upgrade project to use NServiceBus.Transport.AzureServiceBus
4. Configure SubscriptionNameShortener (and RuleNameShortener) to do same job as previous SanitizationStrategy (replace Alpha to A, Beta to B, Gamma to G)

**Expected behavior**
Subscription name using NServiceBus.Transport.AzureServiceBus should be "A.B.G.Endpoint"

**Versions:**
 - NuGet package: 1.0.0
 - OS: Windows 10 build 17134
 - .NET Version: .NET Core 2.2.0

**Additional context**
"Alpha.Beta.Gamma.Endpoint" (is less than 50 symbols) so subscription shortener is not triggered. Thus it provides inconsistent behavior between versions.

Currently SubscriptionShortener is triggered only if name is longer than 50 symbols. One of the solution is either be able to change condition under which logic is triggered or to have a toggle to force shortener

**Comment by SeanFeldman** [comment]
Thank you @dv42 for reporting this.

It's important to note that NServiceBus.Transport.AzureServiceBus is a brand new transport and is not a continuation of NServiceBus.Azure.Transports.WindowsAzureServiceBus transport. Saying that, we have strived to design the transport to be compatible with the legacy ASB transport as much as possible.

If I understand correctly, you have used the legacy transport [sanitization feature](https://docs.particular.net/transports/azure-service-bus/legacy/sanitization) and have sanitized subscription names even though it was well under the maximum number of characters. When converted the endpoint to use the new transport, couldn't get the same subscription name as you used to have.

My question is - if you migrate your endpoints from the old transport to the new one, both your publisher and subscribers will be working using the full ("Alpha.Beta.Gamma.Endpoint") name, wouldn't they? Is the issue with the fact that the names are no longer adhere to your convention? Thank you.

**Comment by dv42** [comment]
Yes, you understand problem correctly. We have sanitized subscription names that were shorter than 50 symbols.

In case if we create brand new endpoint - everything works just fine. However, as far as I can tell, there is a bigger problem then just a convention:

Correct me if I'm wrong, but Endpoint configuration is responsible not only for creating subscription, but for creating a queue as well.
So previously, `new EndpointConfiguration("Alpha.Beta.Gamma.Endpoint")` created subscription "A.B.G.Endpoint" and a queue "A.B.G.Endpoint".
Now, with a new transport, same line of code created subscription "Alpha.Beta.Gamma.Endpoint" and a new queue "Alpha.Beta.Gamma.Endpoint".

Problem that I see:
1. In case if a new message was published to a "A.B.G.Endpoint" queue during update, it won't be processed as soon as updated service will operate with messages in queue "Alpha.Beta.Gamma.Endpoint"
2. After upgrade, "A.B.G.Endpoint" would be still alive, thus every message will be duplicated (one will go to "Alpha.Beta.Gamma.Endpoint" queue and will be processed and a copy of it will go to "A.B.G.Endpoint" queue)

**Potential workaround**
alter endpoint name to `new EndpointConfiguration("A.B.G.Endpoint")`. In this case Subscription and queue name will match previous logic. In this case, ServiceInsight will be affected and it will show both endpoints (with full and with short name). However I'm not sure whether retries will be affected by this (either via ServicePulse or via retry through ServiceInsight)

**Comment by SeanFeldman** [comment]
> alter endpoint name to new EndpointConfiguration("A.B.G.Endpoint"). In this case Subscription and queue name will match previous logic. In this case, ServiceInsight will be affected and it will show both endpoints (with full and with short name). However I'm not sure whether retries will be affected by this (either via ServicePulse or via retry through ServiceInsight).

Retries won't be affected as your queue name won't change. When it comes to ServiceInsight showing endpoints, it will show the short version. If the long-version name is what you see today, you'd have to remove that one from SerciceControl.

**Comment by SeanFeldman** [comment]
@dv42 have you had a chance to review my comments? Let me know your thoughts and wherever you have opened a support case or not. Thank you.

**Comment by dv42** [comment]
@SeanFeldman  Thanks for your replies.
We would test the workaround section in our testing environment but it looks promising on local machine so no need to create a support case (and we can close this issue as well).

Regarding whether it's a valid idea to enforce shortening: it's not necessary for us as long as we have a way to migrate everything to different transport without breaking change and as you mentioned, retries using ServiceInsight/ServicePulse won't be affected.

Just in case, we didn't use sanitization to apply some conventions. Some of our endpoints (or to be fair, event names) were longer than 50 characters, so in order to fit under 50 character limit we applied sanitization for everything (for events as it was mandatory, for endpoints it wasn't). 

So for us, 'A.B.G.Endpoint' is less human readable than full version. It might be useful to provide ability to specify human readable name for endpoint inside constructor or via some property. So it will look like 
`new EndpointConfiguration(queueName = 'A.B.G.Endpoint', serviceInisghtAlias = 'Alpha.Beta.Gamma.Endpoint')`, but this is optional.

**Comment by SeanFeldman** [comment]
Glad to hear that you're moving forward w/o the need in a new API @dv42.

The idea of having a human readable endpoint name that's used in Particular Platform tools is something that we're discussing. There's an [issue](https://github.com/Particular/ServiceControl/issues/1409) in Service Control that talks about that. You might want to add your comment to share your thoughts and experience.

I'll keep the issue opened until you can confirm that your tests are successful and you can go into production. Thank you.

**Comment by SeanFeldman** [comment]
@dv42 looks like the need in this configuration API is not required and the option discussed above works well for you. Could you please confirm that? Thank you 🙂

**Comment by dv42** [comment]
Yes, we've tested this option in our testing environment and decided to proceed with it as we didn't find any major problems, so I believe we can close this ticket.

The only problem that we've found was that ServiceControl on version earlier than 3 (our environments were on v1.43) was unable to process the message from new transport and it wasn't the part of Backward Compatibility guide. However, after infrastructure upgrade to latest ServiceControl everything works as expected (@SeanFeldman you actually helped us with upgrade through Chat support).

**Comment by SeanFeldman** [comment]
@dv42 thank you for the feedback. I've went ahead and have updated the [migration documentation](https://docs.particular.net/transports/azure-service-bus/legacy/migration#additional-considerations) based on your feedback to call out explicitly two things:

1. Not all configuration settings are available in the new transport
1. SC v3 and later is required

Thank you for the confirmation and the valuable feedback.
